File: claude_exp/xv4/shared/ims_connect.py
import struct
from shared.framing import _recv_exact

IRM_HEADER_LEN = 28
PING_TRANSCODE = "PING0001".encode("cp500")


def to_ebcdic(s: str, length: int) -> bytes:
    encoded = s.encode("cp500")
    if len(encoded) >= length:
        return encoded[:length]
    return encoded.ljust(length, b"\x40")  # EBCDIC space


def build_frame(irm_f0, irm_id, client_id, mti=None, data=b"", transcode=None) -> bytes:
    if transcode is None and data:
        transcode = to_ebcdic("TRAN" + (mti or ""), 8)

    payload = (
        struct.pack(">H", IRM_HEADER_LEN)  # 2B
        + b"\x04"                           # 1B
        + bytes([irm_f0])                   # 1B
        + irm_id                            # 8B EBCDIC
        + b"\x00\x00\x00\x00"              # IRM_NAK_RSNCDE(2) + IRM_RES(2)
        + b"\x00\x15\x10\x01"              # IRM_F5, IRM_TIMER, IRM_SOCT, IRM_ES
        + client_id                         # 8B EBCDIC
    )
    if data:
        payload += (transcode or b"\x40" * 8) + data

    frame = struct.pack(">I", len(payload)) + payload
    return frame


def write_response(sock, data: bytes) -> None:
    sock.sendall(struct.pack(">I", len(data)) + data)


def read_response(sock) -> bytes:
    raw = _recv_exact(sock, 4)
    length = struct.unpack(">I", raw)[0]
    return _recv_exact(sock, length)


def read_request(sock) -> tuple:
    raw = _recv_exact(sock, 4)
    length = struct.unpack(">I", raw)[0]
    payload = _recv_exact(sock, length)

    # IRM header: 2B header_len + 1B 0x04 + 1B irm_f0 + 8B irm_id + 4B zeros + 4B flags + 8B client_id
    irm_f0 = payload[3]
    irm_id_bytes = payload[4:12]
    client_id_bytes = payload[20:28]

    remaining = payload[28:]
    if remaining:
        transcode_bytes = remaining[:8]
        iso_data_bytes = remaining[8:]
    else:
        transcode_bytes = b""
        iso_data_bytes = b""

    return irm_f0, client_id_bytes, transcode_bytes, iso_data_bytes

File: claude_exp/routers/router_cpp/src/shared/ims_connect.cpp
#include "shared/ims_connect.h"

#include <algorithm>

#include "shared/ebcdic.h"
#include "shared/framing.h"

namespace xv6::shared {

const std::vector<uint8_t> PING_TRANSCODE = to_ebcdic("PING0001", 8);

namespace {

void append_be16(std::vector<uint8_t>& out, uint16_t v) {
    out.push_back(static_cast<uint8_t>(v >> 8));
    out.push_back(static_cast<uint8_t>(v));
}

void append_be32(std::vector<uint8_t>& out, uint32_t v) {
    out.push_back(static_cast<uint8_t>(v >> 24));
    out.push_back(static_cast<uint8_t>(v >> 16));
    out.push_back(static_cast<uint8_t>(v >> 8));
    out.push_back(static_cast<uint8_t>(v));
}

uint32_t read_be32(const uint8_t* p) {
    return (static_cast<uint32_t>(p[0]) << 24) | (static_cast<uint32_t>(p[1]) << 16) |
           (static_cast<uint32_t>(p[2]) << 8) | static_cast<uint32_t>(p[3]);
}

}  // namespace

std::vector<uint8_t> build_frame(int irm_f0, const std::vector<uint8_t>& irm_id,
                                  const std::vector<uint8_t>& client_id, const std::string& mti,
                                  const std::vector<uint8_t>& data, std::vector<uint8_t> transcode) {
    if (transcode.empty() && !data.empty()) {
        transcode = to_ebcdic("TRAN" + mti, 8);
    }

    std::vector<uint8_t> header;
    header.reserve(IRM_HEADER_LEN);
    append_be16(header, static_cast<uint16_t>(IRM_HEADER_LEN));
    header.push_back(0x04);
    header.push_back(static_cast<uint8_t>(irm_f0));
    header.insert(header.end(), irm_id.begin(), irm_id.end());
    header.insert(header.end(), {0x00, 0x00, 0x00, 0x00});  // IRM_NAK_RSNCDE(2) + IRM_RES(2)
    header.insert(header.end(), {0x00, 0x15, 0x10, 0x01});  // IRM_F5, IRM_TIMER, IRM_SOCT, IRM_ES
    header.insert(header.end(), client_id.begin(), client_id.end());

    std::vector<uint8_t> trailer;
    if (!data.empty()) {
        trailer.insert(trailer.end(), transcode.begin(), transcode.end());
        trailer.insert(trailer.end(), data.begin(), data.end());
    }

    std::vector<uint8_t> out;
    out.reserve(4 + header.size() + trailer.size());
    append_be32(out, static_cast<uint32_t>(header.size() + trailer.size()));
    out.insert(out.end(), header.begin(), header.end());
    out.insert(out.end(), trailer.begin(), trailer.end());
    return out;
}

void write_response(int fd, const std::vector<uint8_t>& data) {
    std::vector<uint8_t> out;
    out.reserve(4 + data.size());
    append_be32(out, static_cast<uint32_t>(data.size()));
    out.insert(out.end(), data.begin(), data.end());
    send_exact(fd, out);
}

std::vector<uint8_t> read_response(int fd) {
    auto len_bytes = recv_exact(fd, 4);
    uint32_t len = read_be32(len_bytes.data());
    return recv_exact(fd, len);
}

ImsRequest read_request(int fd) {
    auto len_bytes = recv_exact(fd, 4);
    uint32_t payload_len = read_be32(len_bytes.data());
    auto payload = recv_exact(fd, payload_len);

    ImsRequest req;
    req.irm_f0 = payload.size() > 3 ? payload[3] : 0;
    if (payload.size() >= 28) {
        req.client_id.assign(payload.begin() + 20, payload.begin() + 28);
    }
    if (payload.size() > IRM_HEADER_LEN) {
        size_t trailer_len = payload.size() - IRM_HEADER_LEN;
        size_t transcode_len = std::min<size_t>(8, trailer_len);
        req.transcode.assign(payload.begin() + IRM_HEADER_LEN,
                              payload.begin() + IRM_HEADER_LEN + transcode_len);
        if (trailer_len > 8) {
            req.iso_data.assign(payload.begin() + IRM_HEADER_LEN + 8, payload.end());
        }
    }
    return req;
}

}  // namespace xv6::shared
